### PR TITLE
docs: Clarify how to switch to Next auth middleware

### PR DESCRIPTION
**Pull request by pohy**
Closes #<issue>

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I performed a functional test on my final commit

---

## Changelog

After switching to JWT strategy the app started crashing. The docs weren't clear enough. I ended up using a debugger to resolve the issue.  
Decided to _clarify_ the docs a bit and add an example.

---

💯

### CODE REVIEW

File: www/src/pages/en/usage/next-auth.md [hunk 1]
```diff
@@ -176,7 +176,30 @@ If for example, you'd like to add a `role` to the `User` model, you would need t
 
 ## Usage with Next.js middleware
 
-Usage of NextAuth.js with Next.js middleware [requires the use of the JWT session strategy](https://next-auth.js.org/configuration/nextjs#caveats) for authentication. This is because the middleware is only able to access the session cookie if it is a JWT. By default, Create T3 App is configured to use the **default** database strategy, in combination with Prisma as the database adapter.
+Usage of NextAuth.js with Next.js middleware [requires the use of the JWT session strategy](https://next-auth.js.org/configuration/nextjs#caveats) for authentication. This is because the middleware is only able to access the session cookie if it is a JWT. By default, Create T3 App is configured to use the **default** database strategy, in combination with Prisma as the database adapter.  
+After switching to the JWT session strategy. Make sure to update the `session` callback in your `[...nextauth].ts`.
```
Comment (joachimjusth) [inline www/src/pages/en/usage/next-auth.md:180]: It should not be `src/server/auth.ts` ?
In new version of the create-t3-app, all the auth configuration stuff is outside of the `[...nextauth].ts`

Comment (pohy) [reply www/src/pages/en/usage/next-auth.md:180]: Sure! Wow, I started using T3 app just a week ago. 😅 
```suggestion
After switching to the JWT session strategy. Make sure to update the `session` callback in your `src/server/auth.ts`.  
```

File: www/src/pages/en/usage/next-auth.md [hunk 2]
```diff
@@ -176,7 +176,30 @@ If for example, you'd like to add a `role` to the `User` model, you would need t
 
 ## Usage with Next.js middleware
 
-Usage of NextAuth.js with Next.js middleware [requires the use of the JWT session strategy](https://next-auth.js.org/configuration/nextjs#caveats) for authentication. This is because the middleware is only able to access the session cookie if it is a JWT. By default, Create T3 App is configured to use the **default** database strategy, in combination with Prisma as the database adapter.
+Usage of NextAuth.js with Next.js middleware [requires the use of the JWT session strategy](https://next-auth.js.org/configuration/nextjs#caveats) for authentication. This is because the middleware is only able to access the session cookie if it is a JWT. By default, Create T3 App is configured to use the **default** database strategy, in combination with Prisma as the database adapter.  
+After switching to the JWT session strategy. Make sure to update the `session` callback in your `[...nextauth].ts`.  
+The `user` object will be `undefined`. Instead, retrieve the user's ID from the `token` object.  
+I.e.:
+
+```diff:pages/api/[...nextauth].ts
```
Comment (pohy) [inline www/src/pages/en/usage/next-auth.md:184]:
```suggestion
```diff:server/auth.ts```

**Review by c-ehrlich — COMMENTED** [review]
This really shouldn't be our job to document - it's specific to NextAuth, not related to the integration of the packages. If the documentation on this is inadequate in the NextAuth documentation, it should be explained better there IMO.

File: www/src/pages/en/usage/next-auth.md [hunk 3]
```diff
@@ -196,3 +225,4 @@ Usage of NextAuth.js with Next.js middleware [requires the use of the JWT sessio
 | NextAuth.js Docs                  | https://next-auth.js.org/               |
 | NextAuth.js GitHub                | https://github.com/nextauthjs/next-auth |
 | tRPC Kitchen Sink - with NextAuth | https://kitchen-sink.trpc.io/next-auth  |
+```
```
Comment (juliusmarminge) [inline www/src/pages/en/usage/next-auth.md]:
```suggestion
```

File: www/src/pages/en/usage/next-auth.md [hunk 4]
```diff
@@ -176,7 +176,36 @@ If for example, you'd like to add a `role` to the `User` model, you would need t
 
 ## Usage with Next.js middleware
 
-Usage of NextAuth.js with Next.js middleware [requires the use of the JWT session strategy](https://next-auth.js.org/configuration/nextjs#caveats) for authentication. This is because the middleware is only able to access the session cookie if it is a JWT. By default, Create T3 App is configured to use the **default** database strategy, in combination with Prisma as the database adapter.
+Usage of NextAuth.js with Next.js middleware [requires the use of the JWT session strategy](https://next-auth.js.org/configuration/nextjs#caveats) for authentication. This is because the middleware is only able to access the session cookie if it is a JWT. By default, Create T3 App is configured to use the **default** database strategy, in combination with Prisma as the database adapter.  
+After switching to the JWT session strategy. Make sure to update the `session` callback in `src/server/auth.ts`.  
+The `user` object will be `undefined`. Instead, retrieve the user's ID from the `token` object.  
+I.e.:
+
+`diff:server/auth.ts`
```
Comment (juliusmarminge) [inline www/src/pages/en/usage/next-auth.md:184]:
```suggestion
```diff:server/auth.ts
```

**Comment by juliusmarminge** [comment]
> This really shouldn't be our job to document - it's specific to NextAuth, not related to the integration of the packages. If the documentation on this is inadequate in the NextAuth documentation, it should be explained better there IMO.

iirc, Balazs don't approve of using the user.id as we do. We tried getting them to make that a default in next-auth but he mentioned some reasons why he didn't like it.

IMO, we could have a section for this but it should be clear that it's not a recommended approach or something 🤔

File: www/src/pages/en/usage/next-auth.md [hunk 5]
```diff
@@ -176,7 +176,36 @@ If for example, you'd like to add a `role` to the `User` model, you would need t
...
+*       session.user.id = user.id;
+
+-     if (session.user && token.sub) {
+-       session.user.id = token.sub;
+        }
+        return session;
+      }
+  },
+  };
+
```
Comment (juliusmarminge) [inline www/src/pages/en/usage/next-auth.md:207]:
```suggestion
  export const authOptions: NextAuthOptions = {
+   session: {
+     strategy: "jwt",
+   },
    callbacks: {
-     session({ session, user }) {
+     session({ session, token }) {
-       if (session.user) {
+       if (session.user && token.sub) {
-         session.user.id = user.id;
+         session.user.id = token.sub;
        }
        return session;
      }
    },
  };
```

**Pull request merged**